kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/MessagePublishContext.java

```diff
@@ -54,16 +58,32 @@ public void completed(Exception exception, long ledgerId, long entryId) {
 
             topic.recordAddLatency(System.nanoTime() - startTimeNs, TimeUnit.MICROSECONDS);
 
-            // In asynchronously send mode, there's a possibility that some messages' offsets are larger than the actual
-            // offsets. For example, two messages are sent concurrently, if the 1st message is completed while the 2nd
-            // message is already persisted, `MessageIdUtils.getCurrentOffset` will return the 2nd message's offset.
-            final long baseOffset = MessageIdUtils.getCurrentOffset(managedLedger) - (numberOfMessages - 1);
+            if (offset < 0) {
+                if (offset != -1L) {
+                    log.error("offset was changed to {} (< 0) but no exception was thrown", offset);
+                }
+                offset = MessageIdUtils.getCurrentOffset(managedLedger);
+            }
+
+            final long baseOffset = offset - (numberOfMessages - 1);
             offsetFuture.complete(baseOffset);
         }
 
         recycle();
     }
 
+    @Override
+    public void setMetadataFromEntryData(ByteBuf entryData) {
+        if (entryData == null) {
+            // When the managed ledger was closed, the `entryData` is null.
+            return;
+        }
+        final BrokerEntryMetadata metadata = Commands.peekBrokerEntryMetadataIfExist(entryData);
+        if (metadata != null && metadata.hasIndex()) {
+            offset = metadata.getIndex();
+        }
```

Comment:
Should we add some log, or return some error for the else path?

Reply:
I added the logs in `completed` to handle the case that `BrokerEntryData` is null or has no index for the specific entry, PTAL again.